cluster/images/etcd/migrate/options.go

```diff
@@ -0,0 +1,242 @@
+	flags.StringVar(&opts.etcdDataPrefix, "etcd-data-prefix", "", "etcd key prefix under which all objects are kept")
+	flags.StringVar(&opts.ttlKeysDirectory, "ttl-keys-directory", "", "etcd key prefix under which all keys with TTLs are kept. Defaults to {etcd-data-prefix}/events")
+	flags.StringVar(&opts.initialCluster, "initial-cluster", "", "comma separated list of name=endpoint pairs. Defaults to etcd-{hostname}=http://localhost:2380")
+	flags.StringVar(&opts.targetVersion, "target-version", "", "version of etcd to migrate to. Format must be '<major>.<minor>.<patch>'")
+	flags.StringVar(&opts.targetStorage, "target-storage", "", "storage version of etcd to migrate to, one of: etcd2, etcd3")
+	flags.StringVar(&opts.etcdServerArgs, "etcd-server-extra-args", "", "additional etcd server args for starting etcd servers during migration steps, --peer-* TLS cert flags should be added for etcd clusters with more than 1 member that use mutual TLS for peer communication.")
+}
+
+func lookupEnv(env string) (string, error) {
+	result, ok := os.LookupEnv(env)
+	if !ok || len(result) == 0 {
+		return result, fmt.Errorf("%s variable unset - expected failure", env)
+	}
+	return result, nil
+}
+
+func lookupEnvOrFail(env string) string {
+	result, err := lookupEnv(env)
+	if err != nil {
+		klog.Fatalf("%v", err)
+	}
+	return result
+}
+
+func defaultName() string {
+	if etcdName, err := lookupEnv(etcdNameEnv); err == nil {
+		return etcdName
+	}
+	klog.Infof("%s variable unset - falling back to etcd-%s variable", etcdNameEnv, etcdHostnameEnv)
+	if etcdHostname, err := lookupEnv(etcdHostnameEnv); err == nil {
+		return fmt.Sprintf("etcd-%s", etcdHostname)
+	}
+	klog.Infof("%s variable unset - falling back to etcd-%s variable", etcdHostnameEnv, hostnameEnv)
+	if hostname, err := lookupEnv(hostnameEnv); err == nil {
+		return fmt.Sprintf("etcd-%s", hostname)
+	}
+	klog.Infof("%s variable unset - falling back to etcd-$(hostname)", etcdHostnameEnv)
+	if hostname, err := os.Hostname(); err == nil {
+		return fmt.Sprintf("etcd-%s", hostname)
+	}
+	klog.Fatalf("Couldn't determine hostname")
+	return "THIS SHOULDN'T HAPPEN"
+}
+
+func (opts *migrateOpts) defaultAndValidateOrFail() {
+	if opts.name == "" {
+		klog.Infof("--name unset - falling back to %s variable", etcdNameEnv)
+		opts.name = defaultName()
+	}
+
+	if opts.dataDir == "" {
+		klog.Infof("--data-dir unset - falling back to %s variable", dataDirEnv)
+		opts.dataDir = lookupEnvOrFail(dataDirEnv)
+	}
+
+	etcdEventsRE := regexp.MustCompile("event")
+	if opts.port == 0 {
+		if etcdEventsRE.MatchString(opts.dataDir) {
+			opts.port = 18631
+		} else {
+			opts.port = 18629
+		}
+		klog.Infof("--port unset - defaulting to %d", opts.port)
+	}
+	if opts.peerPort == 0 {
+		if etcdEventsRE.MatchString(opts.dataDir) {
+			opts.peerPort = 2381
+		} else {
+			opts.peerPort = 2380
+		}
+		klog.Infof("--peer-port unset - defaulting to %d", opts.peerPort)
+	}
+
+	if opts.initialCluster == "" {
```

Comment:
I'm okay with this approach I think. the fallback checks here to the env vars is a bit more verbose than I'd like (and benefits to introducing some utility functions for this?), but functionally it looks fine.